roc/roc.go

```diff
@@ -1,170 +1,142 @@
-	caddress, _ := address.PassRef()
-	__ret := C.roc_sender_bind(csender, caddress)
-	__v := (int32)(__ret)
-	return __v
-}
+// ============================================================================
+//                              address.h
+// ============================================================================
 
-// senderConnect function as declared in roc/sender.h:153
-func senderConnect(sender *Sender, kind PortType, proto Protocol, address *Address) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	ckind, _ := (C.roc_port_type)(kind), cgoAllocsUnknown
-	cproto, _ := (C.roc_protocol)(proto), cgoAllocsUnknown
-	caddress, _ := address.PassRef()
-	__ret := C.roc_sender_connect(csender, ckind, cproto, caddress)
-	__v := (int32)(__ret)
-	return __v
-}
+// NewAddress parses the `ip`, `port` and `family` and initializes the Address object
+func NewAddress(family Family, ip string, port int) (*Address, error) {
+	a := new(Address)
+	a.mem = make([]byte, C.sizeof_roc_address)
+	a.raw = (*C.roc_address)(unsafe.Pointer(&a.mem))
 
-// senderWrite function as declared in roc/sender.h:177
-func senderWrite(sender *Sender, frame *frame) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	cframe, _ := frame.PassRef()
-	__ret := C.roc_sender_write(csender, cframe)
-	__v := (int32)(__ret)
-	return __v
-}
-
-// senderClose function as declared in roc/sender.h:192
-func senderClose(sender *Sender) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	__ret := C.roc_sender_close(csender)
-	__v := (int32)(__ret)
-	return __v
+	cfamily := (C.roc_family)(family)
+	ip = safeString(ip)
+	cip := C.CString(ip)
+	cport := (C.int)(port)
+	errCode := C.roc_address_init(a.raw, cfamily, cip, cport)
+	runtime.KeepAlive(a.mem)
+
+	if errCode == 0 {
+		return a, nil
+	}
+	if errCode < 0 {
+		return nil, ErrInvalidArguments
+	}
+	return nil, ErrInvalidApi
+}
+
+func (a *Address) Family() (Family, error) {
+	f := C.roc_address_family(a.raw)
+	family := (Family)(f)
+	if family == AfInvalid {
+		return family, ErrInvalidArguments
+	}
+	return family, nil
+}
+
+func (a *Address) IP() (string, error) {
+	const buflen = 255
+	sIP := make([]byte, buflen)
+	res := C.roc_address_ip(a.raw, (*C.char)(unsafe.Pointer(&sIP)), buflen)
+	if res == nil {
+		return "", ErrInvalidArguments
+	}
+	return C.GoString(res), nil
+}
+
+func (a *Address) Port() (int, error) {
+	res := C.roc_address_port(a.raw)
+	if res < 0 {
+		return (int)(res), ErrInvalidArguments
+	}
+	return (int)(res), nil
+}
+
+// ============================================================================
+//                              context.h
+// ============================================================================
+
+func OpenContext(contextConfig *ContextConfig) (*Context, error) {
+	c := C.roc_context_open((*C.roc_context_config)(unsafe.Pointer(&contextConfig)))
```

Comment:
And here, we can't just cast ContextConfig to roc_context_config. Although logically they have the same fields in the same order, it's not guaranteed that they have the same layout in memory. E.g. they may have different padding and alignment. Actually I'm not even sure that Go compiler guarantees to preserve the field order in memory.

So instead of a cast, we should allocate C.roc_context_config (better on stack) and do a field-by-field copy from Go struct to C struct.

Reply:
this is what `c-for-go` generated for the type `Context`:
```
type Context C.roc_context
```
shoud I rewrite the type as well?

Reply:
> type Context C.roc_context

This is OK, I guess. Both `C.roc_context` and `Context` are opaque types, so I see no problem here. We never dereference pointers to Context in Go code, just pass the pointer as is to the C code.

Reply:
done